### ISSUE TITLE
Cannot add cospend project (cannot find server)

### ISSUE THREAD

**Issue by Narfinger** [InteractionEngineer/PayForMe#68]
I tried everything to add a cospend project into this app but it always reports that it cannot find the server.
The QR code gives me a giant red X and the other ones say that they cannot find the server.
I am using nextcloud on a local server with a self-signed certificate but I also tried with a let's encrypt certificate.
Is there a way to get better error messages?

**Comment by CapitolRadio** [comment]
Similar issue for me. The QR code doesn't register at all though... 
When I try to add project manually, I can't click 'Add Project' button (it is greyed out).

**Comment by derTunnel** [comment]
Same here

**Comment by azertylr** [comment]
hello, same here
how to add a cospend project please?

**Comment by tiptoptom** [comment]
Do you know the app and cospend version? Did the problem start with an update of the app/cospend?

**Comment by tiptoptom** [comment]
Looks like this problem:
https://github.com/julien-nc/cospend-nc/issues/258

**Comment by azertylr** [comment]
I managed to solve this issue this way:
-use a subdomain got nextcloud (nextcloud.domain.tld) and not a subfolder
-copy the link on cospend
-paste it in the app
-put a random value for the password

**Comment by tiptoptom** [comment]
>  -copy the link on cospend

Which link do you mean exactly? Does it look like this?
https://nextcloud.xxx.com/apps/cospend/s/xxx

>  -put a random value for the password

Really a random value? Did you even set a password?

**Comment by azertylr** [comment]
>
> Which link do you mean exactly? Does it look like this? https://nextcloud.xxx.com/apps/cospend/s/xxx
> 
yes that's correct, copy this link in iOS

> 
> Really a random value? Did you even set a password?

then use the "past link" function in PayForMe: it will complete server url & username
and fill the password field with a random value

it worked for me

**Comment by tiptoptom** [comment]
Well, sometimes that seems to work, sometimes it doesn't.

**Comment by maxammann** [comment]
The issue behind this is rate limiting in Nextcloud I believe. I also got some 429 HTTP errors when vising the shared link in my browser.

**Comment by maxammann** [comment]
Setting:

```
'auth.bruteforce.protection.enabled' => false,
```

works around the issue.

**Comment by Hustenbonbon** [comment]
Hi all, thanks for providing the information!
By now I work in a very different setting and it gets more and more complicated to find the time to maintain PayForMe, but all your feedback helps prioritizing it.

I will try to find the reason the Cospend communication is completely broken and fix it as soon as possible. Maybe @julien-nc can point me in the right direction?

**Comment by Hustenbonbon** [comment]
Question to everyone:
Would you be fine with completely giving up the deprecated login method in favor for a nextcloud login?

**Comment by maxammann** [comment]
> Question to everyone:
> Would you be fine with completely giving up the deprecated login method in favor for a nextcloud login?

Do you mean removing link sharing and requiring a login? That would work for me.

**Comment by tiptoptom** [comment]
> Would you be fine with completely giving up the deprecated login method in favor for a nextcloud login?

For me, it's a big advantage that not all my friends who want access to the project have to create an account first. I would therefore like this option to continue to exist (if that's what you meant).

**Comment by azertylr** [comment]
For simplicity, cannot we use the new official sharing link? it doesn't need a password or a nextcloud account.
Like: https://domain.tld/apps/cospend/s/dc19853041ebc04f444293459e2a485ce

The second option would be to work like MoneyBuster with a nextcloud account to retrieve/create new project automatically

**Comment by maxammann** [comment]
> For simplicity, cannot we use the new official sharing link? it doesn't need a password or a nextcloud account.
> Like: https://domain.tld/apps/cospend/s/dc19853041ebc04f444293459e2a485ce

That is what is used right now and there seems to be an issue on how PayForMe uses that. There is also authentication involved and somehow PayForMe is triggering the bruteforce detection.

**Comment by azertylr** [comment]
yes it seems that it is trying to connect in loop, and not only when we validate

**Comment by tiptoptom** [comment]
I think [this is the place](https://github.com/julien-nc/cospend-nc/blob/23fa9d133b3ef54559594b58b78434ce4d2fdb70/lib/Controller/OldApiController.php#L113) where the [error message](https://github.com/julien-nc/cospend-nc/issues/258#issue-2118178166) is defined.

**Comment by tiptoptom** [comment]
@Hustenbonbon any news here? 🙈

**Comment by Solear** [comment]
Yeah, any news here?
The workaround 
'auth.bruteforce.protection.enabled' => false
like mentioned in https://github.com/julien-nc/cospend-nc/issues/258 doenst work :-)
I wanna go away from splitwise

**Comment by saemy** [comment]
I seem to hit the same issue:
When I add a project I see the following log on my server:
```
"GET /index.php/apps/cospend/api/projects/ojef.../no-pass HTTP/1.1" 401
```

However, my project has uppercase letters in it! When I manually access the link with uppercase letters, everything works:
```
"GET /index.php/apps/cospend/api/projects/oJEf.../no-pass HTTP/1.1" 200
```

I think this is the responsible line in code:
https://github.com/mayflower/PayForMe/blob/49d046c70ca638d759c822653892f0881165b9b5/PayForMe/Services/NetworkService.swift#L155

**Comment by Hustenbonbon** [comment]
Hi, there is unfortunately no update nor will one come from my side. I am not working for Mayflower anymore and access for me was removed from this repository and from the apple account of Mayflower. To my knowledge, there is no one else who is picking it up.
After initial positive response by Mayflower on transfering the repository and the app, my last messages were not responded to anymore.

**Comment by InteractionEngineer** [comment]
Hi everyone! This got into a bit of a mess. 🙈
I've looked into the problems described and it really seems there are different things at play. 

1. There are issues with the QR-Code reader (but manually adding a project worked)
2. There are authentication issues, some suggested turning off the brute force protection in Nextcloud/Cospend would resolve this (please, never do that – it's on of the most basic security features) … but at least we can reduce the connection attempts and stop something like this happening in the future.
3. Possibly related to No. 2: Around September of 2024, Cospend changed their project name hashing to something that also uses uppercase letters, but PayForMe still expects the old style of just lowercased letters (and numbers).
4. @azertylr I'm not sure how your domain structure looked before. If the problem still persists, please open a separate issue  with concrete examples how to reproduce. 

Extra: The call of @Narfinger for better user feedback / error messages is legitimate – I'll keep this in mind as an important task for the future. More improvements in this regard are planned.

**Comment by InteractionEngineer** [comment]
Alright! 😄
As said, I think many of the reported issues have (several) common causes. 

- For now, I think I've fixed No. 1 and No. 3.
- I don't know yet what might have caused No. 2 and will be looking into this at #81.
- I'd like to do the same for No. 4 at #82.

As the reports are some years old now, please provide further evidence ( **= help** ) that those are still problems that should be taken care of. And update to the newest version of PayForMe in the AppStore, available after Apple's review.
